components/statistics/SankeyChartBox/SankeyChartBox.helper.ts

```diff
@@ -0,0 +1,106 @@
+import { BlockchainPathMap } from 'types';
+import {
+  SankeyDataType,
+  SankeyLabelsType,
+  SankeyPriorityType,
+} from './SankeyChartBox.type';
+import { BlockchainMeta } from 'types/meta';
+
+const MAX_NUMBER_OF_SOURCE = 10;
+const MAX_NUMBER_OF_DESTINATION = 10;
+export const DEFAULT_BLOCKCHAIN_COLOR = '#B7D8FF';
+export const SPLIT_SOURCE_DESTINATION = '__';
+
+export const getSankeyChartData = (
+  topPath: BlockchainPathMap[],
+  blockchainDataMap: Map<string, BlockchainMeta>,
+) => {
+  const sourceTotalValueMap = new Map<string, number>();
+  const destinationTotalValueMap = new Map<string, number>();
+  const labels: SankeyLabelsType = {};
+  const priority: SankeyPriorityType = {};
+
+  topPath.forEach((topItem) => {
+    const { key, value } = topItem;
+    const { source, destination } = key;
+    if (source) {
+      const sum = sourceTotalValueMap.get(source) || 0;
+      sourceTotalValueMap.set(source, sum + value);
+    }
+
+    if (destination) {
+      const sum = destinationTotalValueMap.get(destination) || 0;
+      destinationTotalValueMap.set(destination, sum + value);
+    }
+  });
+
+  const topSources = Array.from(sourceTotalValueMap)
+    .map(([source, value]) => ({
+      source,
+      value,
+    }))
+    .sort((a, b) => b.value - a.value)
+    .slice(0, MAX_NUMBER_OF_SOURCE)
+    .map((topItem) => topItem.source);
+
+  const topDestination = Array.from(destinationTotalValueMap)
+    .map(([destination, value]) => ({
+      destination,
+      value,
+    }))
+    .sort((a, b) => b.value - a.value)
+    .slice(0, MAX_NUMBER_OF_DESTINATION)
+    .map((topItem) => topItem.destination);
+
+  const topPathSorted = topPath.sort((a, b) => b.value - a.value);
+
+  const data: SankeyDataType[] = [];
+  let priorityStep = 0;
+  topSources.forEach((sourceItem) => {
+    const topPathBySource = topPathSorted
+      .filter(
+        (pathItem) =>
+          pathItem.key.source === sourceItem &&
+          topDestination.includes(pathItem.key.destination),
+      )
+      .slice(0, MAX_NUMBER_OF_DESTINATION);
+
+    // add SPLIT_SOURCE_DESTINATION to the end of source key to prevent loop in chart
+    const sourceKey = sourceItem + SPLIT_SOURCE_DESTINATION + 'source';
```

Comment:
Could we have some functions to prevent duplicate codes?

Reply:
Because this function is customized for the Sankey chart,
It is not useful anywhere else.